include: "rules/common.smk"

##### Target rules #####

rule all:
    input:
        intermediate_qc=ancient("/".join([output_directory, samplename+".intermediate.db"])),
        gene_expression="/".join([output_directory, samplename+".gene.tsv"]),
        rnaseq="/".join([output_directory, samplename+".rnaseq_metrics.txt"]),
        isoform_expression="/".join([output_directory, samplename+".tx.tsv"]),
        junctions="/".join([output_directory, samplename+".junctions.tsv"]),
        gemini= "/".join([output_directory, samplename+".gemini.db"]),
        bcftools= "/".join([output_directory, samplename+".bcftools.stats"]),
        idx_stats="/".join([output_directory, samplename+".idxstats"]),
        bam="/".join([output_directory, samplename+".mdup.bam"]),
        gatk_pre="/".join([output_directory, samplename+".recal_data.table"]),
        gatk_post="/".join([output_directory, samplename+".post_recal_data.table"]),
        ribo_int=config["resources"]["ribo_int"],
        refflat=config["resources"]["refflat"],
        gtex_db=config["resources"]["gtex_db"]
    params:
        picard=config["executables"]["picard"],
        ngenes=config["script_parameters"]["refflat_ngenes"]
    output:
        "/".join([output_directory, samplename+".rnaseq_results.db"])
    threads: 1
    shell:
         """
         python3 scripts/finalize.py --gene_expression {input.gene_expression} --isoform_expression {input.isoform_expression} \
            --junctions {input.junctions} --gemini {input.gemini} --bcftools {input.bcftools} \
            --gatk_pre {input.gatk_pre} --gatk_post {input.gatk_post} --intermediate {input.intermediate_qc} \
            --idx_stats {input.idx_stats} --bam {input.bam}  \
            --ribo_int {input.ribo_int} --refflat {input.refflat} --gtex_db {input.gtex_db} --picard {params.picard} --ngenes {params.ngenes} \
            --analysis_path {output_directory} \
            --rnaseq {input.rnaseq}
            
         rm -rf {output_directory}/{samplename}.insert_metrics.pdf \
            {output_directory}/{samplename}Log.out {output_directory}/{samplename}Log.progress.out \
            {output_directory}/{samplename}_STARgenome {output_directory}/{samplename}_STARtmp \
            {output_directory}/{samplename}_STARpass1 {output_directory}/{samplename}.stat  \
            {output_directory}/{samplename}_R1_fastqc.zip {output_directory}/{samplename}_R2_fastqc.zip \
            {output_directory}/{samplename}.split.bai {output_directory}/{samplename}.split.bam.bai \
            {output_directory}/{samplename}.recal_reads.bai {output_directory}/{samplename}.filtered.vcf
         """

##### Modules #####

include: "rules/alignment_quantification.smk"
include: "rules/expression_outlier.smk"
include: "rules/junctions.smk"
include: "rules/qc_metrics.smk"
include: "rules/variants_calling.smk"
